### PR TITLE
Add zfs_prepare_disk script for disk firmware install

### PR DESCRIPTION
**Pull request by tonyhutter**
### Motivation and Context
Add a way to update disk firmware before including a disk in pool.

### Description
Have libzfs call a special `zfs_prepare_disk` script before a disk is included into the pool.  The user can edit this script to add things like a disk firmware update or a disk health check.  Use of the script is totally optional. See the `zfs_prepare_disk` manpage for full details.

This also reformats the code so that both `zfs_prepare_disk` and the `zpool.d/` scripts use similar codepaths.

### How Has This Been Tested?
Created a pool on a VM and saw script run.  Added a disk and saw script get run.  Replaced a disk and saw script get run.  Tested autoreplace.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: lib/libzfs/os/linux/libzfs_pool_os.c [hunk 1]
```diff
@@ -338,3 +338,193 @@ zpool_label_disk(libzfs_handle_t *hdl, zpool_handle_t *zhp, const char *name)
...
+
+	(void) nvlist_lookup_string(vdev_nv, ZPOOL_CONFIG_PATH, &path);
+	(void) nvlist_lookup_string(vdev_nv, ZPOOL_CONFIG_VDEV_ENC_SYSFS_PATH,
+	    &enc_sysfs_path);
+
+	upath = zfs_get_underlying_path(path);
+	pool_name = zhp ? zpool_get_name(zhp) : NULL;
+
+	env = zpool_vdev_script_alloc_env(pool_name, path, upath,
+	    enc_sysfs_path, "VDEV_PREPARE", prepare_str);
```
Comment (behlendorf) [inline lib/libzfs/os/linux/libzfs_pool_os.c:475]: It looks like we can unconditionally free `upath` here after it's been used above.

File: lib/libzfs/os/linux/libzfs_pool_os.c [hunk 2]
```diff
@@ -338,3 +338,193 @@ zpool_label_disk(libzfs_handle_t *hdl, zpool_handle_t *zhp, const char *name)
...
+
+	zpool_vdev_script_free_env(env);
+
+	return (rc);
+}
+
+/*
+ * Optionally run a script and then label a disk.  The script can be used to
+ * prepare a disk for inclusion into the pool.  For example, it might update
+ * the disk's firmware or check it's health.
```
Comment (behlendorf) [inline lib/libzfs/os/linux/libzfs_pool_os.c:493]:
```suggestion
 * the disk's firmware or check its health.
```

File: man/man8/zfs_prepare_disk.8 [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+It can be modified by the user to run whatever commands are necessary to prepare
+a disk for inclusion into the pool.
+For example, users can add lines to
+.Nm zfs_prepare_disk
+to do things like update the drive's firmware or check the drive's health.
+.Nm zfs_prepare_disk
+is optional and can be removed if not needed.
+libzfs will look for
+.Nm zfs_prepare_disk
+in LIBEXECDIR (default /usr/local/libexec/zfs/zfs_prepare_disk).
```
Comment (behlendorf) [inline man/man8/zfs_prepare_disk.8:34]: By converting this man page to a `zfs_prepare_disk.8.in` we can dynamically generate these paths from the configure information.  Take a look at the use of `@zfsexecdir@` in `zfs-mount-generator.8.in` for an example.

File: man/man8/zfs_prepare_disk.8 [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+.
+.Ss Properties
+.Nm zfs_prepare_disk
+will be passed the following env variables:
```
Comment (behlendorf) [inline man/man8/zfs_prepare_disk.8:38]:
```suggestion
will be passed the following environment variables:
```

File: man/man8/zfs_prepare_disk.8 [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+.sp
+.Bl -tag -compact -width "org.openzfs.systemd:required-by=unit[ unit]…"
```
Comment (behlendorf) [inline man/man8/zfs_prepare_disk.8:40]: Appears to be unrelated and can be dropped.

File: man/man8/zfs_prepare_disk.8 [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+.Nm POOL_NAME
+is blank at pool creation time, for example.
+.Sh ENVIRONMENT
+.Nm zfs_prepare_disk
+runs with a limited $PATH.
+.Sh EXIT STATUS
+.Nm zfs_prepare_disk
+should return 0 on success, non-zero otherwise.
+If non-zero is returned, the disk will not be included in the pool.
+.
```
Comment (behlendorf) [inline man/man8/zfs_prepare_disk.8:72]: Stray `.`?

**Comment by tonyhutter** [comment]
@behlendorf I included your fixes in my latest push, however I see some FreeBSD build errors and still need to deal with those

File: scripts/zfs_prepare_disk [hunk 7]
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
+#
+# This is an optional helper script that is automatically called by libzfs
+# before a disk is about to be added into the pool.  It can be modified by
+# the user to run whatever commands are necessary to prepare a disk for
+# inclusion into the pool.  For example, users can add lines to this
+# script to do things like update the drive's firmware or check the drive's
+# health.  The script is optional and can be removed if it is not needed.
+#
+# See the zfs_prepare_disk man page for details.
```
Comment (behlendorf) [inline scripts/zfs_prepare_disk:10]:
```suggestion
# See the zfs_prepare_disk(8) man page for details.
```

**Review by behlendorf — APPROVED** [review]
